src/dbnode/encoding/proto/docs/unmarshal.md

```diff
@@ -0,0 +1,50 @@
+# Top Level Scalar Unmarshaler
+
+## Overview
+
+It's recommended that readers familiarize themselves with the [proto3 encoding documentation](https://developers.google.com/protocol-buffers/docs/encoding) before reading the remainder of this document.
+
+The Encoder in this package is responsible for accepting a stream of marshaled Protobuf messages and encoding them into a new compressed stream.
+
+In order to accomplish this, it needs to unmarshal the Protobuf messages so that it can re-encode their values.
+
+Since the schemas for the Protobuf messages are provided dynamically (and thus efficient unmarshaling code can not be generated ahead of time) the easiest way to accomplish the unmarshaling is to rely on a dynamic Protobuf package like `jhump/protoreflect` to perform the heavy lfting.
+
+This is inefficient primarily because unmarshaling into a `*dynamic.Message` is very expensive since it involves a lot of `interface{}` magic as well as allocating a large number of very short-lived objects that are difficult to reuse.
+It's especially inefficient for Protobuf schemas that are optimized for this package (specifically those that make heavy use of top-level scalar fields where allocating objects on the heap just to wrap primitive types is especially wasteful).
+
+As a result, this package implements a `customFieldUnmarshaler` that accepts a `[]byte` which constitutes a marshaled Protobuf message and exposes methods for unmarshaling the top-level scalar fields (i.e the fields that the encoder can perform custom compression on) in an efficient and reusable manner such that in the general case there are no allocations.
+In addition, it uses the `jhump/protoreflect` library to unmarshal any other fields that it can't unmarshal efficiently (although this has zero overhead if those fields are not present).
+
+
+## Implementation
+
+### Overview
+
+The implementation is broken into two parts:
+
+1. The `Buffer` which is almost identical to https://github.com/jhump/protoreflect/blob/master/dynamic/codec.go and provides an interface for iterating a marshaled protobuf message one `<fieldNumber, wiretype, value>` tuple at a time.
+
+2. The `customFieldUnmarshaler` which wraps the `Buffer` and exposes an interface for efficiently unmarshaling top-level scalar fields with no allocations, as well as a fallback mechanism that relies on the `jhump/protoreflect` library for any fields that don't have an efficient unmarshaling codepath (`maps`, `repeated` fields, and nested messages, etc.)
```

Comment:
nit:

> The `customFieldUnmarshaler`, which wraps `buffer`

also, move the period outside the parenthesis: "nested messages, etc.)" -> "nested messages, etc)."